atomics/include/desul/atomics/Generic.hpp

```diff
@@ -126,22 +126,27 @@ atomic_fetch_oper(const Oper& op,
                   typename std::enable_if<sizeof(T) == 4, const T>::type val,
                   MemoryOrder order,
                   MemoryScope scope) {
-  union U {
-    int32_t i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
+  using fixed_width_integer_type = int32_t;
+  static_assert(sizeof(T) == sizeof(fixed_width_integer_type), "");
 
-  oldval.t = *dest;
+  T assume;
+  T newval;
+  T oldval{*dest};
 
   do {
-    assume.i = oldval.i;
-    newval.t = op.apply(assume.t, val);
-    oldval.i = desul::atomic_compare_exchange(
-        (int32_t*)dest, assume.i, newval.i, order, scope);
-  } while (assume.i != oldval.i);
+    assume = oldval;
+    newval = op.apply(assume, val);
+    reinterpret_cast<fixed_width_integer_type&>(oldval) =
+        desul::atomic_compare_exchange(
+            reinterpret_cast<fixed_width_integer_type* const>(dest),
+            reinterpret_cast<fixed_width_integer_type const&>(assume),
+            reinterpret_cast<fixed_width_integer_type const&>(newval),
+            order,
+            scope);
+  } while (reinterpret_cast<fixed_width_integer_type const&>(assume) !=
+           reinterpret_cast<fixed_width_integer_type const&>(oldval));
```

Comment:
Here I was tempted to just do `assume != oldval` but this adds a requirement on `T`